File: algorithm/bit/single_number.cpp
// https://leetcode.com/problems/single-number

/*
Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.
You must implement a solution with a linear runtime complexity and use only constant extra space.

Ex1:
Input: nums = [2,2,1]
Output: 1

Ex2:
Input: nums = [4,1,2,1,2]
Output: 4

Ex3:
Input: nums = [1]
Output: 1

*/

#include <cassert>
#include <vector>
#include <iostream>

using namespace std;

// XOR, the number xor itself is 0.
// Time: O(n), Space: O(1)
int singleNumber(vector<int>& nums) {
    int n = nums.size();
    int res = 0;
    for (int i = 0; i < n; i++) {
        res ^= nums[i];
    }
    return res;
}

int main() {
    vector<int> nums = {2,2,1};
    assert(singleNumber(nums) == 1);
    nums = {4,1,2,1,2};
    assert(singleNumber(nums) == 4);
    nums = {1};
    assert(singleNumber(nums) == 1);
    return 0;
}

File: algorithm/bit/python/single_number.py
"""
Single Number (https://leetcode.com/problems/single-number)

Given a non-empty array of integers nums, every element appears twice except
for one. Find that single one. Must use linear time and constant extra space.

Approaches:
    1. singleNumber - XOR fold across the array. Time: O(n), Space: O(1).
"""

from typing import List


# XOR, the number xor itself is 0.
# Time: O(n), Space: O(1)
def singleNumber(nums: List[int]) -> int:
    res = 0
    for x in nums:
        res ^= x
    return res


if __name__ == "__main__":
    assert singleNumber([2, 2, 1]) == 1
    assert singleNumber([4, 1, 2, 1, 2]) == 4
    assert singleNumber([1]) == 1
